File: backend-go/internal/jobs/scheduler.go
package jobs

import (
	"context"
	"time"

	"github.com/worldsignal/backend/internal/db"
	"github.com/worldsignal/backend/internal/logging"
)

// Scheduler enqueues fetches for sources whose crawl interval has elapsed.
type Scheduler struct {
	db      *db.DB
	workers *Workers
	tick    time.Duration
	log     *logging.Logger
	cancel  context.CancelFunc
	done    chan struct{}
}

// NewScheduler builds a scheduler with the given tick interval.
func NewScheduler(d *db.DB, w *Workers, tick time.Duration) *Scheduler {
	return &Scheduler{db: d, workers: w, tick: tick, log: logging.New("scheduler")}
}

// DueSource is an enabled source eligible for a fetch.
type dueSource struct {
	ID             string
	CrawlFrequency int
	LastFetchedAt  *time.Time
}

// Tick enqueues fetches for due sources. Exported for testing.
func (s *Scheduler) Tick(ctx context.Context, now time.Time) (int, error) {
	rows, err := s.db.Pool.Query(ctx,
		`SELECT "id","crawlFrequency","lastFetchedAt" FROM "Source" WHERE "enabled"=true ORDER BY "priority" ASC`)
	if err != nil {
		return 0, err
	}
	var sources []dueSource
	for rows.Next() {
		var d dueSource
		if err := rows.Scan(&d.ID, &d.CrawlFrequency, &d.LastFetchedAt); err != nil {
			rows.Close()
			return 0, err
		}
		sources = append(sources, d)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return 0, err
	}

	due := 0
	for _, src := range sources {
		var last time.Time
		if src.LastFetchedAt != nil {
			last = *src.LastFetchedAt
		}
		if now.Sub(last) >= time.Duration(src.CrawlFrequency)*time.Second {
			if err := s.workers.EnqueueFetchSource(src.ID); err != nil {
				return due, err
			}
			due++
		}
	}
	if due > 0 {
		s.log.Info("scheduled sources")
	}
	return due, nil
}

// Start runs the scheduler loop (kick once, then on the tick interval).
func (s *Scheduler) Start(parent context.Context) {
	ctx, cancel := context.WithCancel(parent)
	s.cancel = cancel
	s.done = make(chan struct{})
	go func() {
		defer close(s.done)
		if _, err := s.Tick(ctx, time.Now()); err != nil {
			s.log.Error("tick failed", err.Error())
		}
		ticker := time.NewTicker(s.tick)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				if _, err := s.Tick(ctx, time.Now()); err != nil {
					s.log.Error("tick failed", err.Error())
				}
			}
		}
	}()
}

// Stop halts the scheduler loop.
func (s *Scheduler) Stop() {
	if s.cancel != nil {
		s.cancel()
		<-s.done
		s.cancel = nil
	}
}

File: backend/src/jobs/scheduler.ts
import { prisma } from "../db/prisma.js";
import { enqueueFetchSource } from "./workers.js";
import { logger } from "../lib/logger.js";

const log = logger("scheduler");

const TICK_MS = Number(process.env.SCHEDULER_TICK_MS ?? 30_000);
let timer: NodeJS.Timeout | null = null;

/** Enqueue fetches for any enabled source whose crawl interval has elapsed. */
async function tick(): Promise<void> {
  const now = Date.now();
  const sources = await prisma.source.findMany({
    where: { enabled: true },
    select: { id: true, name: true, crawlFrequency: true, lastFetchedAt: true },
    orderBy: { priority: "asc" },
  });

  let due = 0;
  for (const s of sources) {
    const last = s.lastFetchedAt?.getTime() ?? 0;
    if (now - last >= s.crawlFrequency * 1000) {
      await enqueueFetchSource(s.id);
      due++;
    }
  }
  if (due > 0) log.info(`scheduled ${due}/${sources.length} sources`);
}

export function startScheduler(): void {
  if (timer) return;
  // Kick once on boot, then on an interval. singletonKey in enqueue prevents
  // piling up duplicate fetch jobs for the same source.
  void tick().catch((e) => log.error("tick failed", (e as Error).message));
  timer = setInterval(() => {
    void tick().catch((e) => log.error("tick failed", (e as Error).message));
  }, TICK_MS);
  log.info(`scheduler started (tick ${TICK_MS}ms)`);
}

export function stopScheduler(): void {
  if (timer) {
    clearInterval(timer);
    timer = null;
  }
}
